core/kotlinx-coroutines-test/src/TestBuilders.kt

```diff
@@ -0,0 +1,118 @@
+ * conditions.
+ *
+ * Unhandled exceptions thrown by coroutines in the test will be re-thrown at the end of the test.
+ *
+ * @throws UncompletedCoroutinesError If the [testBody] does not complete (or cancel) all coroutines that it launches
+ * (including coroutines suspended on join/await).
+ *
+ * @param context An optional context that MUST contain a [DelayController] and/or [TestCoroutineExceptionHandler]
+ * @param testBody The code of the unit-test.
+ */
+@ExperimentalCoroutinesApi
+fun runBlockingTest(context: CoroutineContext? = null, testBody: suspend TestCoroutineScope.() -> Unit) {
+    val (safeContext, dispatcher) = context.checkArguments()
+    // smart cast dispatcher to expose interface
+    dispatcher as DelayController
+
+    val startingJobs = safeContext.activeJobs()
+
+    val scope = TestCoroutineScope(safeContext)
+    val deferred = scope.async {
+        scope.testBody()
+    }
+    dispatcher.advanceUntilIdle()
+    deferred.getCompletionExceptionOrNull()?.let {
+        throw it
+    }
+    scope.cleanupTestCoroutines()
+    val endingJobs = safeContext.activeJobs()
+    if ((endingJobs - startingJobs).isNotEmpty()) {
+        throw UncompletedCoroutinesError("Test finished with active jobs: $endingJobs")
+    }
+}
+
+private fun CoroutineContext.activeJobs(): Set<Job> {
+    return checkNotNull(this[Job]).children.filter { it.isActive }.toSet()
+}
+
+/**
+ * Convenience method for calling [runBlockingTest] on an existing [TestCoroutineScope].
+ */
+@ExperimentalCoroutinesApi
+fun TestCoroutineScope.runBlockingTest(block: suspend TestCoroutineScope.() -> Unit) {
+    runBlockingTest(coroutineContext, block)
+}
+
+/**
+ * Convenience method for calling [runBlockingTest] on an existing [TestCoroutineDispatcher].
+ */
+@ExperimentalCoroutinesApi
+fun TestCoroutineDispatcher.runBlockingTest(block: suspend TestCoroutineScope.() -> Unit) {
+    runBlockingTest(this, block)
+}
+
+private fun CoroutineContext?.checkArguments(): Pair<CoroutineContext, ContinuationInterceptor> {
+    var safeContext= this ?: TestCoroutineExceptionHandler() + TestCoroutineDispatcher()
+
+    val dispatcher = safeContext[ContinuationInterceptor].run {
+        this?.let {
+            require(this is DelayController) { "Dispatcher must implement DelayController" }
+        }
+        this ?: TestCoroutineDispatcher()
+    }
+
+    val exceptionHandler = safeContext[CoroutineExceptionHandler].run {
+        this?.let {
+            require(this is UncaughtExceptionCaptor) { "coroutineExceptionHandler must implement UncaughtExceptionCaptor" }
+        }
+        this ?: TestCoroutineExceptionHandler()
+    }
+
+    val job = safeContext[Job] ?: SupervisorJob()
+
+    safeContext = safeContext + dispatcher + exceptionHandler + job
+    return Pair(safeContext, dispatcher)
+}
+
+/**
+ * This method is deprecated.
+ *
+ * @see [runBlocking]
+ */
+@Deprecated("This API has been deprecated to integrate with Structured Concurrency.",
```

Comment:
why are we adding a deprecated API ?

Reply:
This is here to aid in automatic refactoring to the new library - the suggested `ReplaceWith` for `TestCoroutineContext` creates code that doesn't have an obvious refactor without knowing the internals of this library.

Not sure if this is a good practice though. Put the code up for more opinions.

Reply:
This is good.